lib/chef/knife/winrm_knife_base.rb

```diff
@@ -0,0 +1,316 @@
+            if authorization_error?(e)
+              unless config[:suppress_auth_failure]
+                # Display errors if the caller hasn't opted to retry
+                ui.error "Failed to authenticate to #{s.host} as #{config[:winrm_user]}"
+                ui.info "Response: #{e.message}"
+                ui.info get_failed_authentication_hint
+                raise e
+              end
+            else
+              raise e
+            end
+          end
+
+          def get_failed_authentication_hint
+            if @session_opts[:basic_auth_only]
+              FAILED_BASIC_HINT
+            else
+              FAILED_NOT_BASIC_HINT
+            end
+          end
+
+          def authorization_error?(exception)
+            exception.is_a?(WinRM::WinRMAuthorizationError) ||
+              exception.message =~ /401/
+          end
+
+          def check_for_errors!
+            @exit_code ||= 0
+            @winrm_sessions.each do |session|
+              session_exit_code = session.exit_code
+              unless success_return_codes.include? session_exit_code.to_i
+                @exit_code = [@exit_code, session_exit_code.to_i].max
+                ui.error "Failed to execute command on #{session.host} return code #{session_exit_code}"
+              end
+            end
+          end
+
+          def success_return_codes
+            # Redundant if the CLI options parsing occurs
+            return [0] unless config[:returns]
+
+            @success_return_codes ||= config[:returns].split(",").collect(&:to_i)
+          end
+
+          def session_from_list
+            @list.each do |item|
+              Chef::Log.debug("Adding #{item}")
+              @session_opts[:host] = item
+              create_winrm_session(@session_opts)
+            end
+          end
+
+          def create_winrm_session(options = {})
+            session = Chef::Knife::WinrmSession.new(options)
+            @winrm_sessions ||= []
+            @winrm_sessions.push(session)
+          end
+
+          def resolve_session_options
+            config[:winrm_port] ||= ( config[:winrm_transport] == "ssl" ) ? "5986" : "5985"
+
+            @session_opts = {
+              user: resolve_winrm_user,
+              password: config[:winrm_password],
+              port: config[:winrm_port],
+              operation_timeout: resolve_winrm_session_timeout,
+              basic_auth_only: resolve_winrm_basic_auth,
+              disable_sspi: resolve_winrm_disable_sspi,
+              transport: resolve_winrm_transport,
+              no_ssl_peer_verification: resolve_no_ssl_peer_verification,
+              ssl_peer_fingerprint: resolve_ssl_peer_fingerprint,
+              shell: config[:winrm_shell],
+              codepage: config[:winrm_codepage],
+            }
+
+            if @session_opts[:user] && (not @session_opts[:password])
+              @session_opts[:password] = config[:winrm_password] = get_password
```

Comment:
I don't see `config[:winrm_password]` being used other than here, this mutation is unfortunate.

Reply:
I made this a lot better. winrm_password was actually set above and then this changed it in a terrible way

Reply:
Yeah some of this kind of stuff you need to be real careful since horribly shitty things like this may actually be contracts between knife superclasses and subclasses like knife-windows.

The side effect of setting those config values, might actually get consumed by the superclass like an API.

The biggest offender of this kind of tightly coupled nonsense was knife-windows, so now that its all going to be together it'll be a lot easier fix both sides of it, but you need to be real careful about inheritance in this mess.